runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -541,6 +573,10 @@ static void *dereferenceStructPointerChain(void *baseStruct, TR::Node *baseNode,
                   }
 
                fieldAddress = TR::Compiler->om.getAddressOfElement(comp, curStruct, offset);
+
+               if (!comp->getSymRefTab()->isImmutableArrayShadow(symRef) &&
+                   isNullValueAtAddress(comp, symRef->getSymbol()->getDataType(), fieldAddress, symRef->getSymbol()))
```

Comment:
If there is a mismatch between the type of the load and the array element type (which can only happen when both are primitive types since `verifyFieldAccess()` has succeeded), or if the address is not naturally aligned, then we have to take special care because it's the array elements that are stable, not just any sequence of bytes that we care to load from within the array

For example, if we were to load a 32-bit integer from a stable array of bytes, and if that 32-bit integer were nonzero, it might still overlap byte elements with value zero that can be overwritten later

As another example, if we were to load a 16-bit integer from a stable array of 16-bit integers, but at an odd address, and if the loaded value were nonzero, it too might still overlap an element with value zero that can be overwritten later

The exact safety test here would be to look at each array element (based on the array element type) that overlaps the value to be loaded, and to give up if any such element is zero

However, I don't think we need that level of complexity (and certainly not at the moment). It should suffice to check that the type of the load (i.e. `symRef->getSymbol()->getDataType()`) is the same as the array element type, and that the address is naturally aligned

Reply:
Yeah, that's a good point. This is another thing that makes stable array handling different from that of array with constant elements! (sorry, maybe I missed that in your earlier comments and also forgot that this type of access can happen). Then, we will need to pass the type of the array element along with the `isBaseStableArray.` Although I still think it would be good to treat them as similar as possible.

Reply:
Yes, you are right. This is another thing that make stable arrays different from arrays with constant elements. Let me see how I can add these checks.

Reply:
@jdmpapin, when you have a chance, could you please point me to a method that returns TR::DataType  for array element given array address?

Reply:
Maybe we can use `J9::ObjectModel::getArrayElementWidthInBytes(TR::Compilation* comp, uintptr_t objectPointer)` and check if sizes of array elements and the symbol reference we are using to access them are the same? In addition we will check (fieldAddress%size ==0 ) ?

Reply:
I think that last suggestion works. We already know from `verifyFieldAccess` that we're either loading a reference from a reference array or a primitive from a primitive array

In the reference case, the sizes will appear to mismatch with compressed references, but it will actually be ok

In the primitive case, the only pairs of distinct types with the same sizes are boolean/byte, short/char, int/float, and long/double. For boolean/byte and short/char, there is no ambiguity about whether a value is zero. For int/float and long/double, we can load an integer element as a floating point number or vice versa. If we load an integer element as floating point, it might appear to be -0.0 and fail to fold even though the element is now constant, but failure to fold is safe. If OTOH we load a floating point element as an integer, we will fold anything that has at least one bit set. The only floating point value that `isNullValueAtAddress()` would have us refuse to fold despite having a bit set is -0.0, but I think it should be safe to fold that value anyway, since it has definitely been modified after the array was zero-initialized

So *as long as it's ok to fold -0.0*, we just require that `fieldAddress` is aligned and the array either is a reference array or has an element size matching the array shadow symbol

Flipping that around, we need to prevent folding if `fieldAddress` is misaligned or the array is a primitive array whose element size does not match the array shadow symbol

Reply:
Right, we might fold -0.0 if we read int as float... But I think it's consistent with JEP's idea of the field/array element becoming constant after second value... Also, Stable is only used inside JCL and unlikely is used in such an ambiguous way.

Reply:
Added the check for size and alignment.